File: Array/10.Majority-Elements/Solution.py
#Solution in O(n) space and O(n) time 
def majority_element(arr):
	arr_length = len(arr)
	count={}
	for i in arr:
		if i in count:
			count[i] +=1
		else:
			count[i] = 1
	for i in count:
		if count[i] > arr_length/2:
			print i
			return
	print "Majority Element does not exist"

arr= [3,3,4,2,4,4,2,4,4]
majority_element(arr)

#Solution in O(n) time

def majority_element_solution(arr):
	count = 1
	maj_index = 0
	for i in range(1,len(arr)):
		if arr[i] ==arr[maj_index]:
			count +=1
		else :
			count -=1
		if count== 0:
			maj_index = i
			count +=1
	if arr.count(arr[maj_index]) > len(arr)/2:
		print arr[maj_index]
	else:
		print "Majority Element does not exist"

arr= [3,3,4,2,4,4,2,4,4]
majority_element_solution(arr)

File: Array/10.Majority-Elements/solution.rb
def majority_element(arr)
  majority_element = arr[0]
  majority_element_count = 0
  arr.each do |e|
    if e == majority_element
      majority_element_count += 1
    else
      majority_element_count -= 1
    end
    if majority_element_count == -1
      majority_element_count = 1
      majority_element = e
    end
  end
  majority_element_count = 0
  arr.each do |e|
    majority_element_count += 1 if e == majority_element
  end
  return majority_element if majority_element_count > arr.length/2
  return nil
end

arr = [3, 3, 4, 2, 4, 4, 2, 4, 4, 4]
p majority_element(arr)
